### PR TITLE
topic4

### PR DESCRIPTION
**Pull request by jordanholtz**

### CODE REVIEW

File: src/API/CartService.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package API;
+
+import java.util.List;
+
+import Model.Item;
+import Model.Product;
+import Model.Transaction;
+import Model.paymentMethods.FinancialPayment;
```
Comment (lucast-globant) [inline src/API/CartService.java:8]: Package names in lower case.

File: src/Model/Cart.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import API.CartService;
+import Model.MailingList.EmailSender;
+import Model.paymentMethods.FinancialPayment;
+
+
+/**
+ * @author Jordan
+ * shopping cart
+ */
+public class Cart implements CartService{
```
Comment (lucast-globant) [inline src/Model/Cart.java:15]: This class name and implementation is wrong. A good name for this class could be CartServiceImpl. CartServiceImpl must handle or manage a Cart.

File: src/Model/Cart.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+	protected List<Item> products;
```
Comment (lucast-globant) [inline src/Model/Cart.java:17]: Remove this and use the Cart you designed and implemented in previous Topics.

File: src/Model/Cart.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	private CustomerService userAccount;
```
Comment (lucast-globant) [inline src/Model/Cart.java:18]: Is this a userAccount or a CustomerService? Please use proper variable names.

File: src/Model/Cart.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	
+	public Cart(CustomerService userAccount){
+		products=new ArrayList<Item>();
+		this.userAccount=userAccount;
+	}
+	
+	public void add(Product prod, int quantity){
+		products.add(new Item(quantity, prod));
```
Comment (lucast-globant) [inline src/Model/Cart.java:26]: Product should be added to user's cart.

File: src/Model/Cart.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	public void cleanCart(){
+		products=new ArrayList<Item>();
+	}
+	
+	/**
+	 * @param category
+	 * @return all items that contains a product with that category
+	 */
+	public List<Item> getProductsByCategory(String category){
+		ArrayList<Item> listByCategory=new ArrayList<Item>();
```
Comment (lucast-globant) [inline src/Model/Cart.java:47]: Products should come from a database or mocked service from previous topics(DataConnection.java).

File: src/Model/Cart.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	 * @param method available payment methods (CREDITCARD, PAYPAL, CASH) 
+	 * @param customer
+	 */
+ 	public Transaction checkout( FinancialPayment method){
+		
+		
+		//begin transaction
+		Transaction t=new Transaction(method, userAccount);				
+
+		EmailSender.getInstance().notifyToList("TRANSACTION_DONE"+"("+t.getId()+")");
```
Comment (lucast-globant) [inline src/Model/Cart.java:68]: Nice!!!

File: src/Model/CustomerAccount.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+import API.CartService;
+import API.UserService;
+import Model.paymentMethods.Cash;
+import Model.paymentMethods.CreditCard;
+import Model.paymentMethods.FinancialPayment;
+import Model.paymentMethods.PaypalAccount;
+
+
+
+public class CustomerAccount implements CustomerService {
```
Comment (lucast-globant) [inline src/Model/CustomerAccount.java:15]: The idea here is to provide e CustomerService that will handle a Customer or a CustomerAccount.

File: src/Model/CustomerAccount.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+	
+	protected String userName;
```
Comment (lucast-globant) [inline src/Model/CustomerAccount.java:17]: You should have a Customer entity with these fields:

protected String userName;
-   private CreditCard creditCard;
-   private PaypalAccount paypal;
-   /**
-    \* money that has in his wallet  
-    */
-   private Cash money;
-   
-   protected Cart myShoppingCart;
-   
-   protected List<CustomerService> friends;
-   protected ProfilePhoto profilePhoto;

File: src/Model/MailingList/EmailSender.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
+package Model.MailingList;
+
+import java.util.ArrayList;
+import java.util.List;
+
+//Subject
+public class EmailSender {
```
Comment (lucast-globant) [inline src/Model/MailingList/EmailSender.java:7]: EmailSender is the Observer, Cart or CartService is the Subject.

File: src/Model/MailingList/testClient.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
+package Model.MailingList;
+
+
+
+/**
+ * @author Jordan
+ * demo showing how to notify observers
+ */
+public class testClient {
```
Comment (lucast-globant) [inline src/Model/MailingList/testClient.java:9]: This class is good for testing purposes, but test code should be placed in scr/test/java source folder as unit test class.

File: src/Model/Users.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import API.UserService;
+
+
+/**
+ * @author Jordan userService's implementation
+ */
+public class Users implements UserService {
```
Comment (lucast-globant) [inline src/Model/Users.java:12]: This class is redundant, you already have a service for Customer. Remove.

File: src/test/_CartService.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+import API.UserService;
+import Model.ConcreteProduct;
+import Model.CustomerAccount;
+import Model.CustomerService;
+import Model.EcommerceService;
+import Model.Product;
+import Model.Transaction;
+
+public class _CartService {
```
Comment (lucast-globant) [inline src/test/_CartService.java:15]: Please use maven for your project. Test classes should go under src/test/java. Assign a good name to test classes, in this case CartServiceImplTest if you are testing CartServiceImpl. If you are unit testing you need to create another test class for each unit (object) you are testing.

File: src/test/_CartService.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+	@Test
+	public void cart_checkout_TransactionDone() {
```
Comment (lucast-globant) [inline src/test/_CartService.java:18]: Rename to cartCheckoutTransactionDoneTest(). Camel case, no underscores and a name that will denote method behavior.

File: src/test/_UserService.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+import org.junit.Test;
+
+import API.UserService;
+import Model.CustomerAccount;
+import Model.CustomerService;
+import Model.UserServiceFactory;
+import Model.Users;
+
+public class _UserService {
```
Comment (lucast-globant) [inline src/test/_UserService.java:13]: Apply the same fixes mentioned in _CartService tests.

**Comment by lucast-globant** [comment]
Please fix.

File: src/Model/Cart.java (hunk 5 above)
Comment (jordanholtz) [inline src/Model/Cart.java:26]: But, this is the user's cart. Each Customer has in her account a reference to her cart (the cart also have a reference to his owner, userAccount). notice that for add a product you need before know the user. Example:
newCustomer1.getCart().add(web.searchProduct("Oferta del Bootcamp"), 1);

File: src/Model/Cart.java (hunk 6 above)
Comment (jordanholtz) [inline src/Model/Cart.java:47]: This is just a consult, this gives you all the products (we are in a cart class) that have a category like the parameter. Why i need to retrieve data from my persistent data, if i already have in memory the actual state. Then when we are done we can persist the actual state.

File: src/Model/Cart.java (hunk 5 above)
Comment (lucast-globant) [inline src/Model/Cart.java:26]: Cart should not implement Cart Service. Cart is the model.

**Pull request closed without merging**